### ISSUE TITLE
About accessing data from the cloud

### ISSUE THREAD

**Issue by csitaula** [rstudio/keras#351]
Is it okay to access the data from the cloud using this command?
data_dir <- gs_local_dir("gs://quarter-deck-529/training-data")
train_data<-file.path(data_dir)
train_generator <- flow_images_from_directory(
  train_data,
  image_data_generator(rescale = 1/255),
  target_size = c(150, 150),
  batch_size = 32
  class_mode = "binary"
)
I tried this code, but raised an error.
train_generator <- flow_images_from_directory(
  gs_local_dir("gs://quarter-deck-529/images/train"),
  image_data_generator(rescale = 1/255),
  target_size = c(150, 150),
  batch_size = 32
  class_mode = "binary"
)
I think former one is correct, though I've not run it on my machine. I tried later one and it failed. It did not recognize the path.

**Comment by csitaula** [comment]
This is the error I got even after importing library(cloudml) on the script.
{
  "message": "could not find function \"gs_local_dir\"",
  "traceback": ["normalize_path(directory)", "flow_images_from_directory(gs_local_dir(\"gs://nyu-training-testing/NYU/train\"), \n    train_datagen, target_size = c(224, 224), batch_size = 20, \n    class_mode = \"categorical\")", "eval(ei, envir)", "eval(ei, envir)", "withVisible(eval(ei, envir))", "source(file = file, local = envir, echo = echo, encoding = encoding)", "withCallingHandlers({\n    source(file = file, local = envir, echo = echo, encoding = encoding)\n    write_run_property(\"completed\", TRUE)\n}, error = function(e) {\n    write_run_metadata(\"error\", list(message = e$message, traceback = capture_stacktrace(sys.calls())))\n    stop(e)\n})", "force(expr)", "tryCatch({\n    tfruns::training_run(file = deploy$entrypoint, context = deploy$context, \n        config = \"cloudml\", flags = deploy$overlay, encoding = \"UTF-8\", \n        echo = TRUE, view = FALSE, run_dir = run_dir)\n}, error = function(e) {\n    training_error <<- e\n})"]
}

**Comment by jjallaire** [comment]
Perhaps you are running an older version of the cloudml package that doesn't have the `gs_local_dir` function? It is most definitely in the package!

**Comment by csitaula** [comment]
This function is already in the cloudml because it recognizes while writing
code. I tried the later code too and it again did not recognize the path.
Can´t I load from my local drive without uploading in the bucket? Which
command I should I use if I have data on C drive?

**Comment by jjallaire** [comment]
If you want to read one set of data locally and another in the cloud you can use a flag as described here: https://tensorflow.rstudio.com/tools/cloudml/articles/storage.html#data-source-configuration

**Comment by csitaula** [comment]
Can´t it possible to load whole data from local drive and manipulate? I
have 400,000 images which is difficult to upload. I am in trouble that if I
don`t use gs_local_dir(), it works locally, but in the cloud, it shows IO
ERROR!

**Comment by jjallaire** [comment]
No, the cloud cannot read from your local drive!

You can use the `gs_copy()` and `gs_rsync()` functions to copy data from your local system to a cloud bucket: https://tensorflow.rstudio.com/tools/cloudml/articles/storage.html#copying-data

**Comment by csitaula** [comment]
Finally,  I solved the problem successfully. 
library(keras)
library(readr)
library(cloudml)
# create the base pre-trained model
base_model <- application_vgg16(weights = 'imagenet', include_top = FALSE)

# model <- keras_model_sequential()%>%
#   base_model%>%
#   layer_flatten() %>%
#   layer_dense(units = 256,activation = 'relu') %>%
#   layer_dense(units = 7,activation = 'sigmoid')
# add our custom layers
predictions <- base_model$output %>% 
  layer_global_average_pooling_2d() %>% 
  layer_dense(units = 1024, activation = 'relu') %>% 
  layer_dense(units = 2, activation = 'softmax')

# this is the model we will train
model <- keras_model(inputs = base_model$input, outputs = predictions)

# first: train only the top layers (which were randomly initialized)
# i.e. freeze all convolutional InceptionV3 layers
freeze_weights(base_model)

# compile the model (should be done *after* setting layers to non-trainable)
model %>% compile(optimizer = 'rmsprop', loss = 'categorical_crossentropy')
train_datagen = image_data_generator(
  rescale = 1/255,
  rotation_range = 40,
  width_shift_range = 0.2,
  height_shift_range = 0.2,
  shear_range = 0.2,
  zoom_range = 0.2,
  horizontal_flip = TRUE,
  fill_mode = "nearest"
)
test_datagen <- image_data_generator(rescale = 1/255)
train_generator <- flow_images_from_directory(
  gs_local_dir("gs://nyu-training-testing/Demo/train/"),
  train_datagen,
  target_size = c(224, 224),
  batch_size = 20,
  class_mode = "categorical"
)
test_generator <- flow_images_from_directory(
  gs_local_dir("gs://nyu-training-testing/Demo/test/"),
  test_datagen,
  target_size = c(224, 224),
  batch_size = 20,
  class_mode = "categorical"
)

model %>% compile(
  loss = "categorical_crossentropy",
  optimizer = optimizer_rmsprop(lr = 2e-5),
  metrics = c("accuracy")
  
)
model%>% fit_generator(
  train_generator,
  steps_per_epoch = 3,
  epochs = 3,
  validation_data = test_generator,
  validation_steps = 5
)
